ocaml/xapi/xapi_xenops.ml

```diff
@@ -2409,6 +2400,17 @@ let maybe_cleanup_vm ~__context ~self =
     Xenopsd_metadata.delete ~__context id;
   end
 
+(* CP-18860: check memory limits if using nested_virt on start *)
+let assert_memory_constraints ~__context ~vm =
+  let platformdata = Db.VM.get_platform ~__context ~self:vm in
+  if Vm_platform.is_true ~key:"nested-virt" ~platformdata ~default:false
```

Comment:
It's a pity that we cannot call `assert_valid_for_current_context` below, and let that function figure out whether we are using nested-virt or not, and what the memory constraints should be. Now we have this bit of policy encoded in two different places (I bet we'll forget one if the policy changes). Then we could simply _always_ check the memory constraints, if not forced, regardless of the `nested-virt` flag (i.e. without this if-statement).

I suppose you didn't do that, because `assert_valid_for_current_context` looks at the metrics rather than the platform key?